cmake/common_compiler_flags.cmake

```diff
@@ -1,94 +1,180 @@
+    set( IS_RELEASE "$<STREQUAL:${TARGET_NAME},template_release>")
+    set( DEBUG_FEATURES "$<OR:$<STREQUAL:${TARGET_NAME},template_debug>,$<STREQUAL:${TARGET_NAME},editor>>" )
+    set( HOT_RELOAD "$<IF:${HOT_RELOAD-UNSET},$<NOT:${IS_RELEASE}>,$<BOOL:${GODOT_USE_HOT_RELOAD}>>" )
+    set( DEBUG_SYMBOLS "$<BOOL:${GODOT_DEBUG_SYMBOLS}>" )
+
+    target_compile_features(${TARGET_NAME}
+            PUBLIC
+            cxx_std_17
+    )
+
+    # These compiler options reflect what is in godot/SConstruct.
+    target_compile_options( ${TARGET_NAME}
+        PUBLIC
+            # Disable exception handling. Godot doesn't use exceptions anywhere, and this
+            # saves around 20% of binary size and very significant build time.
+            $<${DISABLE_EXCEPTIONS}:
+                $<${NOT_MSVC}:-fno-exceptions>
+            >
+            $<$<NOT:${DISABLE_EXCEPTIONS}>:
+                $<${IS_MSVC}:/EHsc>
+            >
+
+            # Enabling Debug Symbols
+            $<${DEBUG_SYMBOLS}:
+                $<${IS_MSVC}: /Zi /FS>
+
+                # Adding dwarf-4 explicitly makes stacktraces work with clang builds,
+                # otherwise addr2line doesn't understand them.
+                $<${NOT_MSVC}:
+                    -gdwarf-4
+                    $<IF:${IS_DEV},-g3,-g2>
+                >
+            >
+
+            $<${IS_DEV}:
+                $<${NOT_MSVC}:-fno-omit-frame-pointer -O0 -g>
+            >
+
+            $<${HOT_RELOAD}:
+                $<${IS_GNU}:-fno-gnu-unique>
+            >
+
+            # TODO Set optimize
+            # "custom" means do nothing and let users set their own optimization flags.
+            #if env.get("is_msvc", False):
+                #if env["optimize"] == "speed":
+                    #env.Append(CCFLAGS=["/O2"])
+                    #env.Append(LINKFLAGS=["/OPT:REF"])
+                #elif env["optimize"] == "speed_trace":
+                    #env.Append(CCFLAGS=["/O2"])
+                    #env.Append(LINKFLAGS=["/OPT:REF", "/OPT:NOICF"])
+                #elif env["optimize"] == "size":
+                    #env.Append(CCFLAGS=["/O1"])
+                    #env.Append(LINKFLAGS=["/OPT:REF"])
+                #elif env["optimize"] == "debug" or env["optimize"] == "none":
+                    #env.Append(CCFLAGS=["/Od"])
+            #else:
+                #if env["optimize"] == "speed":
+                    #env.Append(CCFLAGS=["-O3"])
+                    ## `-O2` is friendlier to debuggers than `-O3`, leading to better crash backtraces.
+                #elif env["optimize"] == "speed_trace":
+                    #env.Append(CCFLAGS=["-O2"])
+                #elif env["optimize"] == "size":
+                    #env.Append(CCFLAGS=["-Os"])
+                #elif env["optimize"] == "debug":
+                    #env.Append(CCFLAGS=["-Og"])
+                #elif env["optimize"] == "none":
+                    #env.Append(CCFLAGS=["-O0"])
```

Comment:
This will either need to be removed (or made to work?) in order for this PR to be mergeable